cookbook/case_studies/feature_engineering/feast_integration/feast_dataobjects.py

```diff
@@ -0,0 +1,107 @@
+from flytekit.configuration import aws
+from datetime import datetime
+import pandas as pd
+import os
+from typing import Type
+from dataclasses import dataclass
+from dataclasses_json import dataclass_json
+from feast import repo_config
+from feast.feature_store import FeatureStore
+from feast.repo_config import RepoConfig
+from flytekit import FlyteContext
+from flytekit.core.type_engine import TypeEngine, TypeTransformer
+from flytekit.models.literals import Literal, Scalar
+from flytekit.models.types import LiteralType, SimpleType
+from feast.infra.offline_stores.file import FileOfflineStoreConfig
+from feast.infra.online_stores.sqlite import SqliteOnlineStoreConfig
+from feast import FeatureStore as FeastFeatureStore
+from google.protobuf.struct_pb2 import Struct
+from google.protobuf.json_format import MessageToDict
+from typing import Any, Dict, Iterable, List, Optional, Tuple, Union
+from feast.entity import Entity
+from feast.feature_view import FeatureView
+from feast.feature_service import FeatureService
+
+
+@dataclass_json
+@dataclass
+class FeatureStoreConfig:
+    registry_path: str
+    project: str
+    s3_bucket: str
+    online_store_path: str = 'online.db'
+
+
+@dataclass_json
+@dataclass
+class FeatureStore:
```

Comment:
do you think this is the best way. we will have to replicate all Feast FeatureStore methods in Flyte?

Reply:
maybe i am missing some point, why not a transformer?

Reply:
On the same lines, overall, the code looks neat! ✨ However, I presume there's boilerplate code—it looks like we are calling the Feast methods twice, although we don't actually do that. Is there a way we can make the user-facing code even simpler? Do we need to have a "Flyte task" per every Feast operation?

Reply:
two very good questions, so let's separate those out. Just to be clear:
1. Why not cut the middleman (in this case these dataclasses) and use a type transformer for the Feast SDK `FeatureStore` object directly? 
    - This way we won't need to maintain these dataclasses at all. This is a great idea.
2. A Flyte Task per Feast operation. 
    - The workflow we're using to showcase the integration, `feast_workflow`, goes through 4 distinct phases: (1) ingesting data into the offline store, (2) loading data from the offline store for training, (3) load data into the online store, and finally (4) retrieve data from the online store for model inference.
    - Having these in separate Flyte Tasks can be used to demonstrate Flyte's strenghts (reproducibility, caching, etc).
    - @samhita-alla , how are you envisioning this simplification? From the user's perspective they only interact with Feast objects once, right? We're doing this translation between dataclasses and the actual Feast `FeatureStore` object, but this will be gone once we have the type transformer.

Reply:
Actually, after putting some thought into this, I don't think we can have a transformer for the `FeatureStore` object because we have custom logic to update online store file. Notice how we upload or download the online store file in each of the methods of the current `FeatureStore` dataclass: [[1]](https://github.com/flyteorg/flytesnacks/blob/4f1222ad83ab450c0af51092daf329df22c8c715/cookbook/case_studies/feature_engineering/feast_integration/feast_dataobjects.py#L68), [[2]](https://github.com/flyteorg/flytesnacks/blob/4f1222ad83ab450c0af51092daf329df22c8c715/cookbook/case_studies/feature_engineering/feast_integration/feast_dataobjects.py#L88-L94), and [[3]](https://github.com/flyteorg/flytesnacks/blob/4f1222ad83ab450c0af51092daf329df22c8c715/cookbook/case_studies/feature_engineering/feast_integration/feast_dataobjects.py#L103).

That said, all this code could be replaced with a type transformer if we could have a remote datasource for `SqliteOnlineStoreConfig`.

Reply:
Feast's registry already does this for us (Feast has [S3RegistryStore](https://github.com/feast-dev/feast/blob/77cdc0e59ba90f2c20e889e89309618d9fef8ae2/sdk/python/feast/infra/aws.py#L154)), notice how we don't have to worry about uploading local changes to the registry to the remote location. 

We could use a similar idea for configuring the sqlite online store. 

That combined with the support for [remote data sources coming in the next release](https://github.com/feast-dev/feast/pull/1789) will remove a lot of the custom used to plug in with Flyte (namely (1) the custom provider, (2) and the dataclass methods used to upload the online store file to s3).

Reply:
@eapolinario How do you think a type transformer implementation could be different from dataclasses? Would that be beneficial in any way?

Concerning the user-facing code, I think having one task per Feast operation helps not think much about Feast-related code as such. The user needs to give the appropriate values, and the rest has to be taken care of by the plugin. This is a bit more complex to implement, but I think it's a deeper integration. 

If we aren't inclined towards making the user-facing code much simpler, this isn't necessary. But if we want to hide some implementation details, this technique could help in doing so. Here's something that I've worked on before to write tasks for the Feast plugin: https://github.com/flyteorg/flytekit/pull/577.

Reply:
> @eapolinario How do you think a type transformer implementation could be different from dataclasses? Would that be beneficial in any way?
@samhita-alla , the idea is to have a type transformer for Feast's [`FeatureStore`](https://rtd.feast.dev/en/latest/#module-feast.feature_store), this way we wouldn't have to worry about exposing the functionality of the main object in our dataclasses. In other words, we'd build the feature store object once and that would be the input of each task that had to interact with Feast.

As for making the user-facing code simpler, I see the appeal of hiding some of this behind a structure like what you proposed in https://github.com/flyteorg/flytekit/pull/577, but unless we can bring more functionality to the plugin, I feel like in its current incarnation it doesn't add a lot of value. I feel like once we can have a type transfomer we can simply expose that object in each task and let the client code deal with the Feast sdk directly.